src/java/main/org/apache/zookeeper/server/DataTree.java

```diff
@@ -478,7 +478,10 @@ public void createNode(final String path, byte data[], List<ACL> acl,
                 HashSet<String> list = ephemerals.get(ephemeralOwner);
                 if (list == null) {
                     list = new HashSet<String>();
-                    ephemerals.put(ephemeralOwner, list);
+                    HashSet<String> _list;
```

Comment:
You can use `computeIfAbsent()` like this:
```java
HashSet<String> list = ephemerals.computeIfAbsent(ephemeralOwner, k -> new HashSet<String>());
```

Reply:
we need synchronization to prevent znode changes while the znode is being snapshot to disk, but for request processing there is only a single thread that does mutations and mutations don't happen while read requests are being processed.